### PR TITLE
Member ACK using tree root

### LINKED ISSUE #791: Member ack should work with the HTTP signature scheme and be over the merkle tree root rather than a nonce
- [ ] fix ack over HTTP and re-enable memberclient test
- [ ] ack should be over root of the tree, rather than nonce, nonce table can be removed etc

### LINKED ISSUE #891: Member ACK support
As it stands, members ACK is not tested in the CI (see https://github.com/microsoft/CCF/blame/master/tests/memberclient.py#L126). We should add support/testing/documentation for this as it is the only way to get a member active.

We really should get this fixed for our next release (v0.8).

### PR DESCRIPTION
**Pull request by jumaffre**
Fixes #791 
Fixes #891 

This PR adds support for member `ack`, using the root of the state tree instead of a random nonce. To do so, the root of the tree is recorded in the `ccf.signatures` table, for easy access from `memberfrontend.h`. 

I've added some docs to explain the process that members have to follow to add a new member and re-enabled the `memberclient.py` end-to-end test. There's some new Python crypto code for members to sign the root of the tree once retrieved via `updateAckStateDigest`.

### CODE REVIEW

File: sphinx/source/members/adding_member.rst [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    $ curl https://<ccf-node-address>/members/updateAckStateDigest  --cacert networkcert.pem --key new_member_privk.pem --cert new_member_cert.pem --header "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id":0, "method":"members/updateAckStateDigest", "params":{}}'
+    {"commit":57,"global_commit":56,"id":0,"jsonrpc":"2.0","result":[<state_digest>],"term":2}
+
+
+Then, the new member should sign the state digest returned by the ``members/updateAckStateDigest`` command with their private key and issue the signature via the ``members/ack`` command.
+
+.. code-block:: bash
+
+    $ curl https://<ccf-node-address>/members/ack  --cacert networkcert.pem --key new_member_privk.pem --cert new_member_cert.pem --header "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id":0, "method":"members/updateAck", "params":{"sig":"[<signature_over_state_digest>]"}}'
```
Comment (jumaffre) [inline sphinx/source/members/adding_member.rst]: I'm trying to find a way to sign the root via the openssl CLI. It shouldn't block this PR to be merged though.

File: sphinx/source/developers/cryptography.rst [hunk 2]
```diff
@@ -47,6 +47,5 @@ Authenticated encryption in CCF relies on AES256-GCM. Ledger authentication reli
 Public-key certificates, signatures, and ephemeral Diffie-Hellman key exchanges all rely on
 elliptic curves. They can be configured to use one of the following implementations:
 
- * ed25519 from `project Everest <https://project-everest.github.io/>`_.
```
Comment (jumaffre) [inline sphinx/source/developers/cryptography.rst]: `ed25519` is not supported with the current version of mbedtls that we've got. I think it's better to just remove it from our docs altogether.

File: tests/infra/crypto.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        with open(privk_path, "rb") as privk:
+            self.privk = load_pem_private_key(
+                privk.read(), password=None, backend=default_backend()
+            )
+
+    def sign(self, bytes_to_sign, hash_alg):
+        return self.privk.sign(bytes_to_sign, ec.ECDSA(hash_alg))
+
+
+# As per mbedtls md_type_t
```
Comment (jumaffre) [inline tests/infra/crypto.py]: Everything below was moved from `votinghistory.py`.

File: sphinx/source/members/adding_member.rst (hunk 1 above)
Comment (achamayou) [reply sphinx/source/members/adding_member.rst]: Shouldn't this be a scurl.sh?

File: sphinx/source/schemas/updateAckStateDigest_result.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "items": {
+    "maximum": 255,
+    "minimum": 0,
+    "type": "number"
+  },
+  "title": "updateAckStateDigest/result",
```
Comment (achamayou) [inline sphinx/source/schemas/updateAckStateDigest_result.json]: ackStateDigest?

File: src/node/history.h [hunk 5]
```diff
@@ -244,15 +248,17 @@ namespace ccf
 
     std::vector<uint8_t> to_v() const
     {
-      size_t vs =
-        sizeof(index) + sizeof(max_index) + root.SIZE + root.SIZE * path->sz;
+      size_t vs = sizeof(index) + sizeof(max_index) + root.h.size() +
+        root.h.size() * path->sz;
```
Comment (eddyashton) [inline src/node/history.h]: Minor: Can you please bracket `(root.h.size() * path->sz);` while you're here? I spent a while staring at this wondering how it ever worked, but obviously `*` != `+`.

File: src/node/rpc/memberfrontend.h [hunk 6]
```diff
@@ -104,8 +105,9 @@ namespace ccf
            auto new_member_id =
              g.add_member(parsed.cert, parsed.keyshare, MemberStatus::ACCEPTED);
 
-           auto ack = tx.get_view(this->network.member_acks);
-           ack->put(new_member_id, {rng->random(SIZE_NONCE)});
+           auto [ma_view, sig_view] =
+             tx.get_view(this->network.member_acks, this->network.signatures);
+           ma_view->put(new_member_id, {{}, sig_view->get(0)->root});
```
Comment (eddyashton) [inline src/node/rpc/memberfrontend.h]: If I'm reading this correctly, this is also clearing the previous signature? I think we should leave it intact if it exists?

File: src/node/rpc/memberfrontend.h [hunk 7]
```diff
@@ -639,25 +662,29 @@ namespace ccf
...
         {
           args.rpc_ctx->set_response_error(
             jsonrpc::StandardErrorCodes::INVALID_PARAMS,
             "Signature is not valid");
           return;
         }
 
-        MemberAck next_ma{rs.sig, rng->random(SIZE_NONCE)};
-        mas->put(args.caller_id, next_ma);
+        MemberAck next_ma{rs.sig, sig_view->get(0)->root};
```
Comment (eddyashton) [inline src/node/rpc/memberfrontend.h]: Similarly to above - should we be updating the ack digest here? It seems we _could_ be storing these values in separate tables (one is updated by `updateAckStateDigest`, the other by `ack` - they don't need to be in the same struct), but if we're storing them side-by-side then they should at least be a matching pair (ie - a sig, and the root it signs over).

File: src/node/rpc/memberfrontend.h (hunk 6 above)
Comment (eddyashton) [reply src/node/rpc/memberfrontend.h]: Ignore this, I wasn't reading it correctly - this is during member creation, so we're not overwriting an existing signature.

File: sphinx/source/schemas/updateAckStateDigest_result.json (hunk 4 above)
Comment (jumaffre) [reply sphinx/source/schemas/updateAckStateDigest_result.json]: I'd rather keep `updateAckStateDigest` as the operation clearly is a `Write` on the KV and still needs to be followed by the `ack` (as per the new documentation page).

File: src/node/rpc/memberfrontend.h (hunk 7 above)
Comment (jumaffre) [reply src/node/rpc/memberfrontend.h]: As discussed, we now store the full signed request, containing the digest of the state along with the next state digest to-be-signed.

File: sphinx/source/members/adding_member.rst (hunk 1 above)
Comment (jumaffre) [reply sphinx/source/members/adding_member.rst]: Now addressed.

File: src/node/clientsignatures.h [hunk 8]
```diff
@@ -34,55 +42,8 @@ namespace ccf
 
     MSGPACK_DEFINE(sig, req, request_body, md);
   };
+  DECLARE_JSON_TYPE(SignedReq)
+  DECLARE_JSON_REQUIRED_FIELDS(SignedReq, sig, req, request_body, md)
   // this maps client-id to latest SignedReq
   using ClientSignatures = Store::Map<CallerId, SignedReq>;
-
-  inline void to_json(nlohmann::json& j, const SignedReq& sr)
```
Comment (jumaffre) [inline src/node/clientsignatures.h]: This was unused and actually wrong since the FTCP signature over msgpack-serialised request was removed.

**Pull request merged**